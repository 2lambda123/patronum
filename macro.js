/* eslint-disable import/no-extraneous-dependencies */
const { createMacro } = require('babel-plugin-macros');
const { addNamed } = require('@babel/helper-module-imports');
const { default: traverse } = require('@babel/traverse');
const babelPlugin = require('effector/babel-plugin');
const factories = require('./babel-plugin-factories.json');

module.exports = createMacro(patronum, {
  configName: 'patronum',
});

function patronum({
  references,
  state,
  babel,
  config: { importModuleName = 'patronum' } = {},
}) {
  const program = state.file.path;

  Object.keys(references).forEach((referenceName) => {
    const id = addNamed(program, referenceName, importModuleName, {
      nameHint: referenceName,
    });

    references[referenceName].forEach((referencePath) => {
      referencePath.node.name = id.name;
    });
  });

  const instance = babelPlugin(babel, {
    factories: factories.map((name) =>
      name.replace('patronum', importModuleName),
    ),
    noDefaults: true,
  });

  instance.pre();
  traverse(program.parent, instance.visitor, undefined, {
    ...state,
    ...instance,
  });
  instance.post();
}
