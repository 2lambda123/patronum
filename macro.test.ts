import pluginTester from 'babel-plugin-tester';
import factories from './babel-plugin-factories.json';

pluginTester({
  title: 'macro',
  plugin: require('effector/babel-plugin'),
  pluginName: 'effector',
  root: __dirname,
  filename: __filename,
  babelOptions: { filename: __filename, root: __dirname },
  snapshot: true,
  tests: {
    'with attachCreators': {
      code: `import { status } from "patronum/status"; const $status = status({ effect });`,
      pluginOptions: {
        importName: 'patronum/status',
        attachCreators: ['status'],
        noDefaults: true,
        addLoc: true,
      },
    },
    'with fabrics': {
      code: `import { status } from "patronum/status"; const $status = status({ effect });`,
      pluginOptions: {
        factories,
        addLoc: true,
      },
    },
    'with fabrics and root import': {
      code: `import { status } from "patronum"; const $status = status({ effect });`,
      pluginOptions: {
        factories,
        addLoc: true,
      },
    },
  },
});

const macroStatusAndDebounce = `
import { status, debounce } from "./macro";
const $status = status({ effect });
const last = debounce({ source: wow, timeout: 10 });
`;

pluginTester({
  plugin: require('babel-plugin-macros'),
  pluginName: 'macros',
  root: __dirname,
  filename: __filename,
  babelOptions: { filename: __filename },
  snapshot: true,
  tests: {
    'with macro': {
      code: macroStatusAndDebounce,
    },
    'renames import': {
      code: macroStatusAndDebounce,
      pluginOptions: {
        patronum: { importModuleName: '@effector/patronum' },
      },
    },
  },
});
