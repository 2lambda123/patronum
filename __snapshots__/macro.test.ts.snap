// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro with attachCreators: with attachCreators 1`] = `

import { status } from "patronum/status";
const $status = status({ effect });

      ↓ ↓ ↓ ↓ ↓ ↓

var _effectorFileName = '/macro.test.ts';
import { status } from 'patronum/status';
const $status = status({
  ɔ: {
    effect,
  },
  config: {
    loc: {
      file: _effectorFileName,
      line: 2,
      column: 16,
    },
    name: '$status',
    sid: 'j4rpgh',
  },
});


`;

exports[`macro with fabrics and root import: with fabrics and root import 1`] = `

import { status } from "patronum";
const $status = status({ effect });

      ↓ ↓ ↓ ↓ ↓ ↓

var _effectorFileName = '/macro.test.ts';
import { withFactory } from 'effector';
import { status } from 'patronum';
const $status = withFactory({
  sid: 'j4rpgh',
  fn: () =>
    status({
      effect,
    }),
  name: '$status',
  method: 'status',
  loc: {
    file: _effectorFileName,
    line: 2,
    column: 16,
  },
});


`;

exports[`macro with fabrics: with fabrics 1`] = `

import { status } from "patronum/status";
const $status = status({ effect });

      ↓ ↓ ↓ ↓ ↓ ↓

var _effectorFileName = '/macro.test.ts';
import { withFactory } from 'effector';
import { status } from 'patronum/status';
const $status = withFactory({
  sid: 'j4rpgh',
  fn: () =>
    status({
      effect,
    }),
  name: '$status',
  method: 'status',
  loc: {
    file: _effectorFileName,
    line: 2,
    column: 16,
  },
});


`;

exports[`macros import from root and rename import: import from root and rename import 1`] = `

import { status, debounce } from "./macro";
const $status = status({ effect });
const last = debounce({ source: wow, timeout: 10 });

      ↓ ↓ ↓ ↓ ↓ ↓

import { debounce as _debounce } from '@effector/patronum';
import { status as _status } from '@effector/patronum';
import { withFactory } from 'effector';
const $status = withFactory({
  sid: 'j4rpgh',
  fn: () =>
    _status({
      effect,
    }),
});
const last = withFactory({
  sid: '-flsuql',
  fn: () =>
    _debounce({
      source: wow,
      timeout: 10,
    }),
});


`;

exports[`macros import from root: import from root 1`] = `

import { status, debounce } from "./macro";
const $status = status({ effect });
const last = debounce({ source: wow, timeout: 10 });

      ↓ ↓ ↓ ↓ ↓ ↓

import { debounce as _debounce } from 'patronum';
import { status as _status } from 'patronum';
import { withFactory } from 'effector';
const $status = withFactory({
  sid: 'j4rpgh',
  fn: () =>
    _status({
      effect,
    }),
});
const last = withFactory({
  sid: '-flsuql',
  fn: () =>
    _debounce({
      source: wow,
      timeout: 10,
    }),
});


`;

exports[`macros renames import: renames import 1`] = `

import { status, debounce } from "./macro";
const $status = status({ effect });
const last = debounce({ source: wow, timeout: 10 });

      ↓ ↓ ↓ ↓ ↓ ↓

import { debounce as _debounce } from '@effector/patronum/debounce';
import { status as _status } from '@effector/patronum/status';
import { withFactory } from 'effector';
const $status = withFactory({
  sid: 'j4rpgh',
  fn: () =>
    _status({
      effect,
    }),
});
const last = withFactory({
  sid: '-flsuql',
  fn: () =>
    _debounce({
      source: wow,
      timeout: 10,
    }),
});


`;

exports[`macros with macro: with macro 1`] = `

import { status, debounce } from "./macro";
const $status = status({ effect });
const last = debounce({ source: wow, timeout: 10 });

      ↓ ↓ ↓ ↓ ↓ ↓

import { debounce as _debounce } from 'patronum/debounce';
import { status as _status } from 'patronum/status';
import { withFactory } from 'effector';
const $status = withFactory({
  sid: 'j4rpgh',
  fn: () =>
    _status({
      effect,
    }),
});
const last = withFactory({
  sid: '-flsuql',
  fn: () =>
    _debounce({
      source: wow,
      timeout: 10,
    }),
});


`;
